File: 프로그래머스/2/42860. 조이스틱/조이스틱.js
function solution(name) {
    const arr = [...name];
    const len = name.length;
    
    const sum = arr.reduce((acc, cur) => acc + getCount(cur), 0);
    
    if (!arr.includes('A')){
        return sum + len - 1;
    }
    
    let result = len - 1;
    for (let i = 0; i < len; i++){
        let idx = i + 1;
        while (idx < len && arr[idx] === 'A'){
            idx++;
        }
        
        result = Math.min(result, i + (len - idx) * 2, 2 * i + len - idx);
    }
    
    return result + sum;
}

function getCount(char){
    const charNum = char.charCodeAt();
    return Math.min(90 - charNum + 1, charNum - 65);
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
def solution(name):
    alphabet = list('ABCDEFGHIJKLMNOPQRSTUVWXYZ')
    count = 0

    for char in name:
        idx = alphabet.index(char)
        up_down = min(idx, 26 - idx)
        count += up_down

    move = len(name) - 1

    for i in range(len(name)):
        next_idx = i +1

        while next_idx < len(name) and name[next_idx] == 'A':
            next_idx += 1
        length_word = i + len(name) - next_idx + min(i, len(name) - next_idx)
        move = min(move, length_word)

    return count + move